README
Python version of GPDfit software
------------------------------------

Usage:  
------
Start ipython shell and do '%run pype'
(if you don't have ipython you can do 'python -i pype.py')

Then proceed with interactive work e.g. 'plotHERMES(data, fits=[(b, m.values)])'
or 'plotHBCSA(ff, fits=[(b, m.values)])'

See test directory for examples. To run tests, install nosetests

Files
------
Approach.py - classes for theoretical approaches to calculation of observables
Model.py - definition of CFFs and FFs
constants.py - constants
data/  -  directory with data files
Data.py -  classes for representation of data
plots.py  - auxiliary functions for plotting results
publish.py - script that publishes datafiles to www dir
pype.py  - main driver script for experimentation
quadrature.py  - Gaussian quadrature routines
README
test - test units. Use nosetest for elegant testing.
theories.db - database of 'theories' (theory = model + approach)
utils.py  - datafile parsing, dataset plotting and other auxilliary routines

Additional example scripts
--------------------------
ex/KMS11.py  - produces figs for neural network paper KMS11


Changelog

Version 0.9.7 2015-10-07 
	- ADD: Merged pype into gepard 
	- CHG: Moved to git and gitlab.com


Version 0.9.6 2013-08-07 
	- ADD: DVEM coefficient functions (just starting, DVEM not usable yet)
	- ADD: GPD E
	- FIX: All test for "big" NPB paper (2008) fixed

Version 0.9.5 2010-06-25 

	- ADD: Evaluation of gamma function is also controlled by
	       SPEED, saving significant time in fitting
	- ADD: Compiles also with Intel ifort, NAG f95 and GNU gfortran compilers,
	       including OpenMP parallelization.
	- ADD: python extension pygepard.so created
	- ADD: third SO(3) partial wave to 'FIT' model
	       

	{Merged from sid  67:85}
	- ADD: Evolution of GPDs at eta=0 and eta=x
	- ADD: Possibility of eta-dependent ansatz
	- ADD: Mathematica interface completely redone so both
	       calculation of CFFs and fitting is possible from
		   within Mathematica
	- ADD: Evolved Wilson coefficients can be now calculated in
	       the initialization phase which leads to much faster
		   execution
	- FIX: comparison to PEGASUS is OK at NNLO as well
	- FIX: SPEED=3 should be used instead of SPEED=4 now

Version 0.9.3  (2006-12-17)

	- ADD: non-singlet evolution
	- ADD: non-singlet NNLO
	- ADD: MSbar non-diagonal evolution
	- ADD: intelligent choice of y-range for drawing of PDF(x)
	- ADD: many example programs in ex subdirectory which produce
	       data for plots in a big paper
	- ADD: automatic plotting and psfragging of plots
	- FIX: dependence on factorization scale corrected now
	- BUG: SPEED=4 is much more unreliable than before

	{Merged from 0.9-fix2:}
	- FIX: charge-factor bugs (it was not NF-dependent, and was
	       altogher missing from DVCS)
	- ADD: PDF(x), resulting from fit, are also drawn now
	- ADD: Additional experimental data points for fitting

Version 0.9.2  (2006-08-03)

	{Merged from 0.9-fix1:}
	- Fixed F2(P) bug that forced P=1 in DIS case

Version 0.9.1  (2006-07-26)

	- Added possibility of Gaussian integration on 2**ACC points
	  with new parameter ACC=1..6. 
	- Added test routines 'houches' and 'accuracy' for checking 
	  accuracy w.r.t. Les Houches benchmark and w.r.t
	  referent most precise gepard result
	  
Version 0.9  (2006-07-23)

	- Wilson coefficients calculated in initialization phase and
	  together with anomalous dimensions put into common block
	- Integration moved from slatec to 8-point Gaussian
	- Adacf part integrated into source
	- PGPLOT support added
	- Compiles cleanly on both Linux and Cygwin
	- Additional user-friendly documentation in gepard.pdf

Version 0.88  (2006-07-02)

	- Initial version
	- Imported into subversion versioning system
	- Target gepard.exe is broken (so calling from Mathematica
		doesn't work) but "pure Fortran" targets work and produce
		data equivalent with Figs 1 and 2 of Letter

- Version 0.82 (2006-04-06)
	-Numerical integration improved by subdivision of path:
	-LAM..-1/2..1/2..LAM 

- Version 0.8 2006-03-24                                                                                                       

- Version 0.7 2006-03-20                                                                                                       


src/ej.F
C     ****h* gepard/ej.F
C  FILE DESCRIPTION
C    conformal moments of GPD E ansaetze
C
C    $Id:$
C     *******


C     ****s* ej.F/EJ
C  NAME
C     EJ  --  conformal moment of input-scale GPD  E_{J}
C         FCM(1) -- singlet quark
C         FCM(2) -- gluon
C         FCM(3) -- non-siglet quark NS(+)
C         FCM(4) -- non-siglet quark NS(-)
C  DESCRIPTION
C     returns E_{J} for various ansaetze, depending on the variable
C     ANSATZ
C  SYNOPSIS

      SUBROUTINE EJ(J, FCM)

      IMPLICIT NONE
      DOUBLE COMPLEX J, FCM(4) 

C  INPUTS
C           J -- conformal moment
C  OUTPUT
C         FCM -- input scale singlet GPD E_{J}
C  PARENTS
C     INITGPD, GETMBGPD
C  CHILDREN
C     RATGAMMA, CLNGAMMA, CBETA, POCHHAMMER
C  SOURCE
C

      DOUBLE COMPLEX HSEA, HU, HD
      DOUBLE PRECISION NORMS
      DOUBLE COMPLEX CLNGAMMA, CBETA, POCHSEA, POCHG, POCHHAMMER
      DOUBLE COMPLEX JMA(2), CORR(2)
      DOUBLE COMPLEX QJ, BETADIP, BETAEXP
      INTEGER LHBETA(4), LHBETAF(4)
      DOUBLE PRECISION LHA(4), LHLAM(4), ALPT(2)
#include "header.F"
      DATA LHBETA / 3, 4, 6, 5 /
      DATA LHBETAF / 6, 24, 720, 120 /
      DATA LHA / 5.1072D0, 3.06432D0, 0.1939875D0, 1.7D0 /
      DATA LHLAM / -0.8D0, -0.8D0, 0.1D0, 0.1D0 /

*     Evolves with NS(-) so not used for most ansaetze
*     only exception is NSMHOU
      FCM(4) = ZERO
*      'Toy' singlet ansatz
      IF (ANSATZ .EQ. 'TOY') THEN
            FCM(1) = 454760.7514415856 * EXP(CLNGAMMA(0.5d0 + J)) /
     &            EXP(CLNGAMMA(10.6d0 + J))
            FCM(2) = 17.837861981813603 * EXP(CLNGAMMA(-0.1d0 + J)) /
     &            EXP(CLNGAMMA(4.7d0 + J))
            FCM(3) = ZERO
*      'Toy' non-singlet ansatz
      ELSE IF (ANSATZ .EQ. 'NSTOY') THEN
            FCM(1) = ZERO
            FCM(2) = ZERO
            FCM(3) = POCHHAMMER(DBLCMPLX(0.5d0, 0.0d0), 4) /
     &                  POCHHAMMER(DBLCMPLX(0.5d0, 0.0d0) + J, 4)
*  ## ansaetze used in PLB paper hep-ph/0605237
      ELSE IF ((ANSATZ .EQ. 'HARD') .OR. (ANSATZ .EQ. 'SOFT')) THEN
            PAR(12) = 1.1d0
            PAR(13) = 0.25d0
            PAR(23) = 0.25d0
            POCHSEA = (8.0d0, 0.0d0)
            POCHG = (6.0d0, 0.0d0)
            IF (ANSATZ .EQ. 'HARD') THEN
                  PAR(21) = 0.4d0
                  PAR(22) = PAR(12) + 0.1d0
            ELSE IF (ANSATZ .EQ. 'SOFT') THEN
                  PAR(21) = 0.3d0
                  PAR(22) = PAR(12)
            END IF
            PAR(11) = (2.0d0/3.0d0) - PAR(21)
            FCM(1) = PAR(11) * CBETA(
     &      DBLCMPLX(1.0d0 - PAR(12) - PAR(13)*DEL2, 0.0d0) 
     &                                  + J, POCHSEA) / CBETA(
     &            DBLCMPLX(2.0d0 - PAR(12), 0.0d0), POCHSEA) /
     &            (1.0d0 - DEL2)**3
            FCM(2) = PAR(21) * CBETA(    
     &          DBLCMPLX(1.0d0-PAR(22)-PAR(23)*DEL2, 0.0d0) 
     &                                  + J, POCHG) / CBETA(
     &            DBLCMPLX(2.0d0 - PAR(22), 0.0d0), POCHG) /
     &            (1.0d0 - DEL2)**3
            FCM(3) = ZERO
*           Killing E contribution
            PAR(18) = 0.0d0
            PAR(28) = 0.0d0
*  ## ansatz used for testing (programs 'test' and 'accuracy')
      ELSE IF (ANSATZ .EQ. 'TEST') THEN
            FCM(1) = PAR(11) / (1 - DEL2/PAR(14)**2)**3 / POCHHAMMER(
     &            DBLCMPLX(1.0d0 - PAR(12) - PAR(13)*DEL2, 0.0d0)
     &             + J, 8) * POCHHAMMER(
     &            DBLCMPLX(2.0d0 - PAR(12), 0.0d0), 8)
            FCM(2) = PAR(21) / (1 - DEL2/PAR(24)**2)**3 / POCHHAMMER(
     &            DBLCMPLX(1.0d0 - PAR(22) - PAR(23)*DEL2, 0.0d0)
     &             + J, 6) * POCHHAMMER(
     &            DBLCMPLX(2.0d0 - PAR(22), 0.0d0), 6)
            FCM(3) = ZERO
*           Killing E contribution
            PAR(18) = 0.0d0
            PAR(28) = 0.0d0
      ELSE IF (ANSATZ .EQ. 'EFL') THEN
* Use    ENS = NS  (with kappa=KAPS=PAR(18) defining normalization)
          PAR(111) = PAR(11)
*        same for ENG = NG  = 0.6 - NS
          PAR(121) = (0.6d0 - PAR(11))
          FCM(1) = QJ(J, 9, PAR(111), PAR(112), PAR(113), PAR(119)) *
     &                BETADIP(J, PAR(114), PAR(115), PAR(116))
          FCM(2) = QJ(J, 7, PAR(121), PAR(122), PAR(123), PAR(129)) *
     &                BETADIP(J, PAR(124), PAR(125), PAR(126))
          FCM(3) = ZERO
      ELSE IF (ANSATZ .EQ. 'EFLEXP') THEN
* Use    ENS = NS  (with kappa=KAPS=PAR(18) defining normalization)
          PAR(111) = PAR(11)
*        same for ENG = NG  = 0.6 - NS
          PAR(121) = (0.6d0 - PAR(11))
          FCM(1) = QJ(J, 9, PAR(111), PAR(112), PAR(113), PAR(119)) *
     &                BETAEXP(PAR(114))
          FCM(2) = QJ(J, 7, PAR(121), PAR(122), PAR(123), PAR(129)) *
     &                BETAEXP(PAR(124))
          FCM(3) = ZERO
      ELSE IF (ANSATZ .EQ. 'FITNG') THEN
* Don't use first sum-rule constraint (and not that pochs are different!)
          ALPT(1) = PAR(12) + PAR(13)*DEL2
          ALPT(2) = PAR(22) + PAR(23)*DEL2
          JMA(1) = J - ALPT(1)
          JMA(2) = J - ALPT(2)
          CORR(1) = EXP(   CLNGAMMA(DBLCMPLX(7.0d0 + PAR(12), 0.0d0)) -
     &           CLNGAMMA(DBLCMPLX(7.0d0 + ALPT(1), 0.0d0))   )
          CORR(2) = EXP(   CLNGAMMA(DBLCMPLX(7.0d0 + PAR(22), 0.0d0)) -
     &           CLNGAMMA(DBLCMPLX(7.0d0 + ALPT(2), 0.0d0))   )
        IF ( PID .EQ. 0 ) THEN 
*         ---  DIS  ---
          FCM(1) = PAR(11) * POCHHAMMER(DBLCMPLX(2.0d0 - PAR(12),0.0d0) 
     &              , 8) / POCHHAMMER(1.0d0 - PAR(12)
     &          + J , 8) * (1.0d0 - PAR(12) + J) / (1.0d0 + JMA(1))
          FCM(2) = PAR(21) * POCHHAMMER(DBLCMPLX(2.0d0 - PAR(22),0.0d0)
     &              , 6) / POCHHAMMER(1.0d0 - PAR(22)
     &          + J , 6) * (1.0d0 - PAR(22) + J) / (1.0d0 + JMA(2))
          FCM(3) = ZERO
        ELSE  
          FCM(1) = PAR(11) * POCHHAMMER(DBLCMPLX(2.0d0 - PAR(12),0.0d0)
     &              , 8) / POCHHAMMER(1.0d0 - PAR(12)
     &          + J , 8) * (1.0d0 - PAR(12) + J) / (1.0d0 + JMA(1)) /
     &           (1.0d0 - DEL2/(PAR(14)+PAR(15)*J))**PAR(16) *
     &        ( 1.0d0 - PAR(19)*PI*(XI/2.0d0)**(JMA(1)+1.0d0)*SQRT(PI)*
     &         2.0d0**(-2.0d0*ALPT(1))*EXP( CLNGAMMA(1.0d0+J) +
     &          CLNGAMMA(1.0d0+J+ALPT(1)) - CLNGAMMA(1.5d0+J) - 
     &          CLNGAMMA(1.0d0+JMA(1)) - 2.0d0*CLNGAMMA(
     &           DBLCMPLX(ALPT(1), 0.0d0)) ) / SIN(PI*ALPT(1)) *
     &          (1.0d0 - (SIN(PI*ALPT(1)) -
     &           SIN(J*PI))/SIN(PI*JMA(1)) ) * CORR(1)  )
          FCM(2) = PAR(21) * POCHHAMMER(DBLCMPLX(2.0d0 - PAR(22),0.0d0)
     &              , 6) / POCHHAMMER(1.0d0 - PAR(22)
     &          + J , 6) * (1.0d0 - PAR(22) + J) / (1.0d0 + JMA(2)) /
     &           (1.0d0 - DEL2/(PAR(24)+PAR(25)*J))**PAR(26) *
     &        ( 1.0d0 - PAR(29)*PI*(XI/2.0d0)**(JMA(2)+1.0d0)*SQRT(PI)*
     &         2.0d0**(-2.0d0*ALPT(2))*EXP( CLNGAMMA(1.0d0+J) +
     &          CLNGAMMA(1.0d0+J+ALPT(2)) - CLNGAMMA(1.5d0+J) - 
     &          CLNGAMMA(1.0d0+JMA(2)) - 2.0d0*CLNGAMMA(
     &           DBLCMPLX(ALPT(2), 0.0d0)) ) / SIN(PI*ALPT(2)) *
     &          (1.0d0 - (SIN(PI*ALPT(2)) -
     &           SIN(J*PI))/SIN(PI*JMA(2)) ) * CORR(2) )
          FCM(3) = ZERO
        ENDIF
*  ## ansatz used for "big paper"  hep-ph/0703179 
      ELSE IF (ANSATZ .EQ. 'FITBP') THEN
            HU = PAR(31) * POCHHAMMER(DBLCMPLX(1.0d0 - PAR(32), 0.0d0) 
     &              , 4) / POCHHAMMER(DBLCMPLX(1.0d0 - PAR(32), 0.0d0)
     &          + J , 4) * (DBLCMPLX(1.0d0 - PAR(32), 0.0d0) + J) /
     &           (DBLCMPLX(1.0d0 - PAR(32) - PAR(33)*DEL2, 0.0d0) + J) /
     &           (1.0d0 - DEL2/(PAR(34)+PAR(35)*J))**PAR(36)
            HD = PAR(41) * POCHHAMMER(DBLCMPLX(1.0d0 - PAR(42), 0.0d0) 
     &              , 4) / POCHHAMMER(DBLCMPLX(1.0d0 - PAR(42), 0.0d0)
     &          + J , 4) * (DBLCMPLX(1.0d0 - PAR(42), 0.0d0) + J) /
     &           (DBLCMPLX(1.0d0 - PAR(42) - PAR(43)*DEL2, 0.0d0) + J) /
     &           (1.0d0 - DEL2/(PAR(44)+PAR(45)*J))**PAR(46)
* Two options, uncomment only one!
*    1. Take sea normalization as free parameter
            NORMS = PAR(11)
*    2. Constrain sea normalization by momentum sum-rule (it has to
*           be declared as fixed in MINUIT.CMD then).
!            NORMS = 1.0d0 - PAR(21) - 2.0d0*(1.0d0-PAR(32))/(5.0d0 -
!     &               PAR(32)) - (1.0d0-PAR(42))/(5.0d0-PAR(42))
            HSEA = NORMS * POCHHAMMER(DBLCMPLX(2.0d0 - PAR(12), 0.0d0) 
     &              , 8) / POCHHAMMER(DBLCMPLX(1.0d0 - PAR(12), 0.0d0)
     &          + J , 8) * (DBLCMPLX(1.0d0 - PAR(12), 0.0d0) + J) /
     &           (DBLCMPLX(1.0d0 - PAR(12) - PAR(13)*DEL2, 0.0d0) + J) /
     &           (1.0d0 - DEL2/(PAR(14)+PAR(15)*J))**PAR(16)
         FCM(1) = HSEA + HU + HD
         FCM(2) = PAR(21) * POCHHAMMER(DBLCMPLX(2.0d0 - PAR(22), 0.0d0) 
     &              , 6) / POCHHAMMER(DBLCMPLX(1.0d0 - PAR(22), 0.0d0)
     &          + J , 6) * (DBLCMPLX(1.0d0 - PAR(22), 0.0d0) + J) /
     &           (DBLCMPLX(1.0d0 - PAR(22) - PAR(23)*DEL2, 0.0d0) + J) /
     &           (1.0d0 - DEL2/(PAR(24)+PAR(25)*J))**PAR(26)
         FCM(3) = ZERO
*           Killing E contribution
            PAR(18) = 0.0d0
            PAR(28) = 0.0d0
      ELSE IF (ANSATZ .EQ. 'NSFIT') THEN
            FCM(1) = ZERO
            FCM(2) = ZERO
*    Ugly kludge: NS valence = down
            HD = PAR(41) * POCHHAMMER(DBLCMPLX(1.0d0 - PAR(42), 0.0d0) 
     &              , 4) / POCHHAMMER(DBLCMPLX(1.0d0 - PAR(42), 0.0d0)
     &          + J , 4) * (DBLCMPLX(1.0d0 - PAR(42), 0.0d0) + J) /
     &           (DBLCMPLX(1.0d0 - PAR(42) - PAR(43)*DEL2, 0.0d0) + J) /
     &           (1.0d0 - DEL2/(PAR(44)+PAR(45)*J))**PAR(46)
            NORMS = PAR(11)
            HSEA = NORMS * POCHHAMMER(DBLCMPLX(2.0d0 - PAR(12), 0.0d0) 
     &              , 8) / POCHHAMMER(DBLCMPLX(1.0d0 - PAR(12), 0.0d0)
     &          + J , 8) * (DBLCMPLX(1.0d0 - PAR(12), 0.0d0) + J) /
     &           (DBLCMPLX(1.0d0 - PAR(12) - PAR(13)*DEL2, 0.0d0) + J) /
     &           (1.0d0 - DEL2/(PAR(14)+PAR(15)*J))**PAR(16)
          FCM(3) = HD - 0.5d0 / (2.0d0 + 0.5d0) * HSEA
*  ## ansatz for comparison with Les Houches PDF evolution benchmark
      ELSE IF (ANSATZ .EQ. 'HOUCHE') THEN
        FCM(1) =
     &      LHA(1)*LHBETAF(1) / POCHHAMMER(J-DBLCMPLX(LHLAM(1), 0.0D0),
     &        LHBETA(1)+1)
     &  + LHA(2)*LHBETAF(2) / POCHHAMMER(J-DBLCMPLX(LHLAM(2), 0.0D0),
     &        LHBETA(2)+1)
     &  + LHA(3)*LHBETAF(3) / POCHHAMMER(J-DBLCMPLX(LHLAM(3), 0.0D0),
     &        LHBETA(3)+1) * 4.8D0
     &  + LHA(3)*LHBETAF(3) / POCHHAMMER(J+1-DBLCMPLX(LHLAM(3), 0.0D0),
     &        LHBETA(3)+1) * (-2.4D0)
        FCM(2) = 
     &      LHA(4)*LHBETAF(4) / POCHHAMMER(J-DBLCMPLX(LHLAM(4), 0.0D0),
     &        LHBETA(4)+1) 
        FCM(3) = ZERO
        FCM(4) = ZERO
*           Killing E contribution
            PAR(18) = 0.0d0
            PAR(28) = 0.0d0
      END IF

      RETURN
      END
C     ****

